### ISSUE TITLE
TextEditor: value returns HTML with Quill editor specific CSS (aka is PF's Quill implementation useless?)

### ISSUE THREAD

**Issue by firsthour** [primefaces/primefaces#4486]
## 1) Environment
- PrimeFaces version: 6.2.15
- Application Wildfly 12
- Affected browsers: Firefox, Chrome

## 2) Expected behavior

I expected this to work like the old p:editor, for example centering text would have CSS like "text-align: center" and making text big would have CSS like "font-size: 2.5em".

## 3) Actual behavior

p:textEditor's value returns Quill specific CSS classes such as "ql-align-center" and "ql-size-huge". These are useless outside out of the Quill editor as the texteditor.css looks like:

`.ql-editor .ql-size-huge {
  font-size: 2.5em;
}`

When we moved from p:editor to p:textEditor in 6.0.6 a few years ago, I thought it was a straight upgrade. I really don't think they're compatible at all though even though the p:editor showcase page says it is deprecated and to use p:textEditor instead.

p:editor provided straight HTML, p:textEditor provides HTML with editor specific markup that is useless outside the editor.

I spent time this morning digging into Quill, something I probably should have done a while ago but never bothered to (the editor stuff in our tool is barely used, until now, of course). It seems like Quill is really intended for its users to use its Delta format which is a JSON thing that could be converted to XHTML with some work.

Even Primefaces' showcase is broken, if you center align text or make text big and click Submit, those changes are not carried through. There is no center CSS altogether and the Huge text size references the editor CSS.

https://screenshotscdn.firefoxusercontent.com/images/5ec43d69-c668-4cca-aaba-85e8f3f0505d.png

I'm not sure what my suggestion is except that I don't think p:textEditor should be considered a straight upgrade from p:editor. I'm now aware of the PF extensions CKEditor and I will probably move to that. I saw there were licensing issues bringing that into PF proper, and that sounds too bad.

**Comment by tandraschko** [comment]
i don't think we can change something here (if there is no option in quill).
You can also use the pe:ckeditor.

**Comment by firsthour** [comment]
I apologize that I made so many duplicate entries, I have no idea how it happened.

One thing to consider is exposing the Delta object that Quill provides through its API through getContents(). I believe this is the intended way to use Quill, simply getting the inner HTML provides _an_ answer but for lots of functionality within Quill, I don't think it's _the_ answer.

https://quilljs.com/docs/api/#getcontents

**Comment by firsthour** [comment]
I also just edited in the last 3 paragraphs of my issue, I submitted early before which probably started the trigger of even more submissions somehow. Something to note though that the showcase suffers the same problem:

https://screenshotscdn.firefoxusercontent.com/images/5ec43d69-c668-4cca-aaba-85e8f3f0505d.png

**Comment by melloware** [comment]
@firsthour It would be great if you want to deep dive this and figure out of getContents will fix the issue?

**Comment by firsthour** [comment]
Honestly I wrote up to my team this morning that we should just walk away from the Quill implementation, if/when we get some time I will look into pe:ckEditor. Quill's delta markup seems great if you want to later convert to multiple outputs, but we just need HTML. They also provide no out of the box way to convert to HTML and reading their github it seems they did before 1.0.

Leveraging getContents() will probably require an additional dependency. I saw this on NPM, I have no idea if this could be leveraged:

https://www.npmjs.com/package/quill-delta-to-html

Anyways, I don't really think Quill is bad, I just think it should be noted somewhere that not all of the WYSIWYG functions work out of the box. It would be easy enough to add some more CSS to your local file to handle CSS class "ql-size-huge" but that would also mean keeping up with all of Quill's changes/adds when PF updates.

Our CMS-lite thing we have in our application is used so infrequently it took us over 2 years to discover a lot of the stuff didn't work, so I don't really have the money to go after this, or the personal interest for that matter.

**Comment by melloware** [comment]
I agree with you.  I think this issue on Quill is a related complaint: https://github.com/quilljs/quill/issues/2029

**Comment by ggjersund** [comment]
Agree to this. Should have an option for HTML output, otherwise the statement of WYSIWYG is simply false. Also @firsthour, the package you linked to does not convert it into HTML. It also uses custom classes.

**Comment by firsthour** [comment]
That's too bad. We converted to pe:ckEditor in like 3 minutes from p:textEditor. You do need two additional dependencies (PrimeFaces Extensions and something specific to CK Editor) but it's worth it for us.

**Comment by koan00** [comment]
While I agree the transition to Quill has been painful, the HTML from Quill is not useless outside Quill. You just need to include the Quill CSS rules. We do this to render the Quill HTML output as a PDF for our users and it does work. Just adding this info as another option for anyone who might read this.

**Comment by hfluz** [comment]
I agree that the way it is, the textEditor isn't very useful since most of times will want to display the content outside of textEditor component.

Can't textEditor specific styles be added to Primefaces or themes CSS?

As it was already commented, we can easily add the styles to our applications CSS, but it would be harder to keep track of changes between versions.

**Comment by melloware** [comment]
I wouldn't add it to my application.css i would simply include a link to the quill JS CSS on your page.

```html
<link href="https://cdn.quilljs.com/1.3.6/quill.snow.css" rel="stylesheet">
```

**Comment by hfluz** [comment]
> I wouldn't add it to my application.css i would simply include a link to the quill JS CSS on your page.
> 
> ```
> <link href="https://cdn.quilljs.com/1.3.6/quill.snow.css" rel="stylesheet">
> ```

Thanks, I was trying to find their CSS in their Github project, but I didn't have success. I think this solution is good enough, maybe it could be added to Primefaces documentation.

**Comment by tandraschko** [comment]
Will close it for now, it's not related to the PrimeFaces wrapper.

**Comment by selvait90** [comment]
I have added ` class="ql-editor"` to the parent element which helped to apply right font size in the view.
* Generated html code and added ql-editor class to parent element


```
<div class="ql-editor">
	<p>
		<span class="ql-size-small">small </span>
		Normal 
		<span class="ql-size-large">Large</span>
		<span class="ql-size-huge">Huge</span>
	</p>
</div>

```

* style in text editor css 

```

.ql-editor .ql-size-small {
    font-size: .75em;
}

.ql-editor .ql-size-large {
    font-size: 1.5em;
}

.ql-editor .ql-size-huge {
    font-size: 2.5em;
}

```

**Comment by mauromol** [comment]
Could the following example be useful for PF to configure Quill to use inline styles instead of custom classes: https://github.com/quilljs/quill/issues/1093 ?

Embedding Quill CSS in you own application, or either pointing to their CSS, is not a viable solution for all use cases: what if you're using p:textEditor to generate HTML mail bodies? I don't want to put in them references to external CSS files just to make the text in my e-mail centered!

Also, [this page](https://developpaper.com/question/using-vue-quill-editor-to-get-the-html-of-the-input-content-how-to-display-the-same-effect-in-the-external-div-as-in-the-editor/) seems to suggest that adding just the style itself is not enough, because Quill JavaScript is also involved. But honestly I didn't delve into it.

In short, having the recommended rich-text component in PF that produces "proprietary" and non-portable HTML code by default is really bad, IMHO. Especially if this is used for very basic things like text alignment and font size.

**Comment by melloware** [comment]
It's really up to PrimeTek. They use QuillJS in PrimeNG, PrimeReact, PrimeFaces, and PrimeVue.  So its their editor they have chosen.  To me the real flaw lies in how QuillJS works as CKEditor and CLEdit both generated proper HTML without the need of a new stylesheet.
